README.md
# Laravel Query Cache

[![Latest Version on Packagist](https://img.shields.io/packagist/v/juampi92/laravel-query-cache.svg?style=flat-square)](https://packagist.org/packages/juampi92/laravel-query-cache)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/juampi92/laravel-query-cache/run-tests?label=tests)](https://github.com/juampi92/laravel-query-cache/actions?query=workflow%3ATests+branch%3Amaster)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/juampi92/laravel-query-cache/Check%20&%20fix%20styling?label=code%20style)](https://github.com/juampi92/laravel-query-cache/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amaster)
[![Total Downloads](https://img.shields.io/packagist/dt/juampi92/laravel-query-cache.svg?style=flat-square)](https://packagist.org/packages/juampi92/laravel-query-cache)

This package provides a set of macros to cache your Laravel Queries just like Cache::remember.

```php
$featuredPost = Post::published()->orderByMostViews()
    ->cacheDay('post:featured') // <- Here
    ->first();
```

## Installation

You can install the package via composer:

```bash
composer require juampi92/laravel-query-cache
```

That's it! No config or Trait necessary. The package auto-discovery will boot the macros.

## Usage

Instead of doing:

```php
Cache::remember('post:count', $ttl, () => Post::published()->count());
```

You now do:

```php
Post::published()->cache('post:count', $ttl)->count();
```

You can use it in Eloquent Queries as well as in normal Queries.

```php
DB::table('posts')
    ->whereNotNull('published_at')
    ->latest()
    ->cacheHour('post:latest')
    ->first();

Posts::published()
    ->cacheHour('post:count')
    ->count();
```

List of macros:

```php
Post::cache('cache:key', $ttl)->get();
Post::cacheMinute('cache:key')->first();
Post::cacheHour('cache:key')->pluck('id');
Post::cacheDay("cache:key:$id")->find($id);
Post::cacheWeek('cache:key:paginate:10')->paginate(10);
Post::cacheForever('cache:key')->count();
```

## Advanced usage

### Tags

```php
Post::query()
    ->where(...)
    ->cache('post:count')->tags(['posts'])
    ->count();
```

### Different store

```php
Post::query()
    ->where(...)
    ->cache('post:count')->store('redis')
    ->count();
```

### Add your custom cache duration

This is maybe more advanced, but you can do so by [opting out of discovery](https://laravel.com/docs/8.x/packages#opting-out-of-package-discovery), and then importing it yourself:

```php
use Juampi92\LaravelQueryCache\LaravelQueryCacheServiceProvider as BaseServiceProvider;

class LaravelQueryCacheServiceProvider extends BaseServiceProvider
{
    protected function getCustomTimes(): array
    {
        return array_merge(
            parent::getCustomTimes(),
            [
                'rememberForever' => null,
                'cacheFifteenMinutes' => 60 * 15,
            ]
        );
    }
}
```

The original method has
```php
[
    'cacheForever' => null,
    'cacheMinute' => 60,
    'cacheHour' => 60 * 60,
    'cacheDay' => 60 * 60 * 24,
    'cacheWeek' => 60 * 60 * 24 * 7,
]
```

## Disclaimer

This package is supposed to be a nice integration of Cache remember inside the Query Builder.
If you're looking for a advanced eloquent specific cache, I recommend to check out [laravel-eloquent-query-cache](https://github.com/renoki-co/laravel-eloquent-query-cache).

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Juampi92](https://github.com/juampi92)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-query-cache` will be documented in this file.

## 1.0.0 - 202X-XX-XX

- initial release


tests/TestCase.php
<?php

namespace Juampi92\LaravelQueryCache\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Illuminate\Support\Facades\Cache;
use Juampi92\LaravelQueryCache\LaravelQueryCacheServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    public function setUp(): void
    {
        parent::setUp();

        $this->clearCache();

        $this->loadMigrationsFrom(__DIR__.'/stubs/database/migrations');
        $this->artisan('migrate', ['--database' => 'sqlite']);

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Juampi92\\LaravelQueryCache\\Tests\\stubs\\database\\factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelQueryCacheServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        $app['config']->set('database.default', 'sqlite');
        $app['config']->set('database.connections.sqlite', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
        $app['config']->set(
            'cache.driver',
            getenv('CACHE_DRIVER') ?: env('CACHE_DRIVER', 'array')
        );
    }

    /**
     * Clear the cache.
     *
     * @return void
     */
    protected function clearCache()
    {
        Cache::flush();
    }
}


tests/UsageTest.php
<?php

namespace Juampi92\LaravelQueryCache\Tests;

use Juampi92\LaravelQueryCache\Tests\stubs\Post;

class UsageTest extends TestCase
{
    /** @test */
    public function should_throw_runtime_exception_if_used_incorrectly()
    {
        $this->expectException(\RuntimeException::class);

        Post::query()->cacheForever('posts:published:count')->whereNotNull('published_at')->count();
    }

    /** @test */
    public function should_throw_argument_error_when_using_cache_without_ttl()
    {
        $this->expectException(\ArgumentCountError::class);

        Post::query()->cache('posts:published:count')->count();
    }
}


tests/stubs/Post.php
<?php

namespace Juampi92\LaravelQueryCache\Tests\stubs;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Post extends Model
{
    use HasFactory;

    public function scopePublished(Builder $query): void
    {
        $query->whereNotNull('published_at');
    }

    public function scopeWrittenBy(Builder $query, string $username): void
    {
        $query->where('author_username', $username);
    }
}


src/Macros/CacheMacro.php
<?php

namespace Juampi92\LaravelQueryCache\Macros;

use ArgumentCountError;
use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
use Illuminate\Database\Query\Builder as QueryBuilder;
use Juampi92\LaravelQueryCache\CacheHighOrderFunction;

class CacheMacro
{
    /**
     * @return \Closure
     */
    public function __invoke(): \Closure
    {
        /**
         * @param string $key
         * @param \DateTimeInterface|\DateInterval|int $ttl
         * @return CacheHighOrderFunction
         */
        return function (string $key, $ttl = null): CacheHighOrderFunction {
            if (! $ttl) {
                throw new ArgumentCountError(
                    'Too few arguments to function cache(), $ttl was missing. For fixed-length cache times, use `cacheForever`, `cacheMinute`, `cacheHour`, `cacheDay`, `cacheWeek`'
                );
            }

            /** @var QueryBuilder|EloquentBuilder $this */
            return new CacheHighOrderFunction($this, $key, $ttl);
        };
    }

    /**
     * @param \DateTimeInterface|\DateInterval|int|null $ttl
     * @return \Closure
     */
    public static function customTTL($ttl): \Closure
    {
        return function ($key) use ($ttl): CacheHighOrderFunction {
            /** @var QueryBuilder|EloquentBuilder $this */
            return new CacheHighOrderFunction($this, $key, $ttl);
        };
    }
}


src/CacheHighOrderFunction.php
<?php

namespace Juampi92\LaravelQueryCache;

use Illuminate\Cache\TaggedCache;
use Illuminate\Contracts\Cache\Repository;
use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
use Illuminate\Database\Query\Builder as QueryBuilder;
use Illuminate\Support\Facades\Cache;

class CacheHighOrderFunction
{
    private string $key;

    /** @var \DateInterval|\DateTimeInterface|int|null */
    private $ttl;

    private $query;

    private array $tags = [];

    /** @var Repository|TaggedCache */
    private $cache;

    /**
     * CachedQueryBuilder constructor.
     * @param \Illuminate\Database\Query\Builder|\Illuminate\Database\Eloquent\Builder $query
     * @param string $key
     * @param \DateTimeInterface|\DateInterval|int|null $ttl
     */
    public function __construct($query, string $key, $ttl)
    {
        $this->query = $query;
        $this->key = $key;
        $this->ttl = $ttl;
        $this->tags = [];
        $this->cache = Cache::store();
    }

    public function store(string $name): self
    {
        $this->cache = Cache::store($name);

        return $this;
    }

    public function tags(array $names): self
    {
        $this->tags = $names;

        return $this;
    }

    public function __call($method, $arguments)
    {
        if (count($this->tags)) {
            $this->cache = $this->cache->tags($this->tags);
        }

        if (! $this->ttl) {
            return $this->cache->rememberForever(
                $this->key,
                fn () => $this->callQuery($method, $arguments)
            );
        }

        return $this->cache->remember(
            $this->key,
            $this->ttl,
            fn () => $this->callQuery($method, $arguments)
        );
    }

    private function callQuery($method, $arguments)
    {
        $result = $this->query->$method(...$arguments);

        if ($result instanceof EloquentBuilder || $result instanceof QueryBuilder) {
            throw new \RuntimeException('You cant cache a query. Remember to use ->cache() right before executing the query');
        }

        return $result;
    }
}


src/LaravelQueryCacheServiceProvider.php
<?php

namespace Juampi92\LaravelQueryCache;

use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
use Illuminate\Database\Query\Builder as QueryBuilder;
use Illuminate\Support\ServiceProvider;
use Juampi92\LaravelQueryCache\Macros\CacheMacro;

class LaravelQueryCacheServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->registerBaseMacro();

        foreach ($this->getCustomTimes() as $macroName => $ttl) {
            $this->registerCustomMacro($macroName, $ttl);
        }
    }

    protected function getCustomTimes(): array
    {
        return [
            'cacheForever' => null,
            'cacheMinute' => 60,
            'cacheHour' => 60 * 60,
            'cacheDay' => 60 * 60 * 24,
            'cacheWeek' => 60 * 60 * 24 * 7,
        ];
    }

    private function registerBaseMacro(): void
    {
        QueryBuilder::macro('cache', call_user_func(new CacheMacro));
        EloquentBuilder::macro('cache', call_user_func(new CacheMacro));
    }

    /**
     * Will register every custom
     * @param  string  $macroName
     * @param $ttl
     */
    private function registerCustomMacro(string $macroName, $ttl = null): void
    {
        QueryBuilder::macro($macroName, CacheMacro::customTTL($ttl));
        EloquentBuilder::macro($macroName, CacheMacro::customTTL($ttl));
    }
}
